### DISCUSSION TITLE
Improve language recognition for Quarto Projects using R or Python on GitHub

### DISCUSSION THREAD

**Discussion by fBedecarrats** [quarto-dev/quarto-cli#9810] [Q&A] [answered]
## Description
When using Quarto with R or Python, GitHub does not correctly recognize the primary languages of the repository. Instead, it primarily identifies the repository as TeX or HTML, depending on the output format of the rendered files. This misclassification affects the visibility and discoverability of Quarto projects that utilize R or Python.

## Steps to reproduce
1. Create a Quarto project with R or Python code chunks;
2. Render the project to HTML or PDF;
3. Push the project to a GitHub repository.

## Expected behaviour
GitHub only recognizes and displays TeX or HTML as the corresponding languages in the repository, ignoring the R or Python code.

## Proposed solution
This is ultimately a GitHub issue, but I believe that a request from the Quarto development team would be more likely prioritized than a request from a lambda Quarto user. I suggest that a lead developer from the Quarto team bring up this issue to the GitHub development team.

**Comment by mcanouil** [comment]
That's a GitHub issue/feature.
See <https://github.com/github-linguist/linguist>.

You can configure the automatic recognition operated by GitHub.
That's not something that Quarto team can really operate on.
Linguist cannot detect markdown code blocks and there is nothing that can be done on Quarto side.

**Reply by fBedecarrats** [reply]
Exactly. And such issue is much more likely to be taken care of by Github devs if it is submitted by Quarto devs instead of a lambda quarto user. Don't you think that such poor language recognition on Github is detrimental to Quarto visibility?

**Reply by mcanouil** [reply, marked as answer]
That's beyond Quarto.
"qmd" file are detected as markdown.

Linguist does not detect language in markdown code blocks because that's markdown.
That's the same thing with Rmarkdown files.

If you use Python code using reticulate/R in an R script, Python code won't be recognised as Python.

Look at what linguist does and how it works.
Override parameters as you wish and as described in the repository.
No matter who open an issue, it's very unlikely Markdown will be considered as a coding language in GitHub statistics and that code within code will ever be detected by linguist.

Note that the "languages" section on GitHub does not count "markdown" as a coding language (qmd and rmd are markdown).

**Reply by mcanouil** [reply]
Somewhat related:
- https://github.com/quarto-dev/quarto-cli/issues/7437

**Reply by jimjam-slam** [reply]
Did not know you could use `.gitattributes` to override these — thanks, @mcanouil! I generally write analyses in R and then make graphics in OJS, so I would probably configure my projects like:

```
*/index.qmd linguist-language=JavaScript
analysis/*/index.qmd linguist-language=R
```

(Tougher if you like to mix languages in a Quarto doc, but I imagine quite a few folks don't do that too often!)

**Comment by fBedecarrats** [comment]
Hello @mcanouil and many thanks for the feedback. 
I must be honnest, I did not even realize that quarto files were recongized as quarto. I have several projects, and quarto never is recognized by itself:
1. [This one for instance:](https://github.com/BETSAKA/Rural_obs_Madagascar) is recognized by Github as 100% html, besides being generated entirely from quarto
2. [This other quarto project](https://github.com/fBedecarrats/available_data_makay) is quite similar in its structure, but it is recognized by Github as 89% javascript, 9% CSS, and the rest is R and other (because I have a few .R, .csv, .png files in it). Quarto is not recognized for this repo.
3. [This other project is also fully built from quarto](https://github.com/BETSAKA/conservation-deforestation-madagascar), but it is recognized by Github as TEX (99%) and Shell (1%), although there is not a single tex file in it. 

In these 3 example: the github projects contain several long quarto files that are the core of it, plus theur rendered webpages (html, js, css) and in these 3 cases, quarto is never identified by github, and the rendered content is recognized very inconsistently (jss and css not identified in 1., html not identified in 2., Tex hallucinated in 3).
I am in train now, but I will try the fix suggested by @jimjam-slam later.

**Reply by mcanouil** [reply]
> I did not even realize that quarto files were recongized as quarto

Quarto does not exist as a language.
It's defined as similar to Rmarkdown and both are identified as Markdown.
As I said, by default GitHub/Linguist don't recognise Markdown as a programming language thus is not accounted in repository statistics.

> I am in train now, but I will try the fix suggested by @jimjam-slam later.

This is directly from linguist repository I linked. You have a more extensive documentation in there.
There is even an example to override the default about Markdown not accounted.

**Reply by fBedecarrats** [reply]
Thank you @mcanouil for the rectification. I misunderstood the purpose of the #7437 issue you linked to. I think that besides technical documentation for developpers, it would be useful to provide some documentation on these aspects to the users (in the same way that quarto documentation includes [something on github pages for instance](https://quarto.org/docs/publishing/github-pages.html)). I might propose something if I manage to understand how it works.

**Reply by fBedecarrats** [reply]
Please feel free to close the discussion if you think that it is appropriate.

**Reply by mcanouil** [reply]
I don't think setting `.gitattributes` has anything to do with Quarto
and belongs to Quarto documentation. This is 100 % GitHub feature and is very well documented in their documentation and in the linguist repository. Most users won't need this.

GitHub Pages is completely different because it's a web host service very used by GitHub users.
Note that even for this, not all are documented in Quarto documentation. For instance, you won't find the new "deploy from action" feature. (The documentation is even a bit outdated as you can see from the few opened issues on the subject).

If you want advanced configuration of your repository, you'll have to look on GitHub documentation and so will the users interested in advanced use of GitHub.

**Reply by eitsupi** [reply]
Maybe you want to see this github-linguist/linguist#2756

IIUC, this is a GitHub design decision.

**Reply by fBedecarrats** [reply]
Thanks @eitsupi !
